File: WinForms/VB/WMTSManager/WMTSForm.vb
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports TatukGIS.NDK
Imports TatukGIS.NDK.WinForms
Imports TatukGIS.RTL

Namespace WMTSManager

    ''' <summary>
    ''' Summary description for WMTSForm.
    ''' </summary>
    Public Class WMTSForm
        Inherits System.Windows.Forms.Form

        Private lServers As Label

        Private cbxServers As ComboBox

        Private WithEvents btnConnect As Button

        Private lLayers As Label

        Private cbxLayers As ComboBox

        Private cbInvertAxis As CheckBox

        Private WithEvents btnAdd As Button

        Private tkn As TGIS_Tokenizer

        Private GIS As TGIS_ViewerWnd

        Public Function getGIS() As TGIS_ViewerWnd
            Return GIS
        End Function

        Public Sub setGIS(ByVal _gis As TGIS_ViewerWnd)
            GIS = _gis
        End Sub

        ''' <summary>
        ''' Required designer variable.
        ''' </summary>
        Private components As System.ComponentModel.IContainer

        Public Sub New()
            MyBase.New
            '
            ' Required for Windows Form Designer support
            '
            InitializeComponent()
            '
            ' TODO: Add any constructor code after InitializeComponent call
            '
        End Sub

        ''' <summary>
        ''' Clean up any resources being used.
        ''' </summary>
        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            If disposing Then
                If (Not (components) Is Nothing) Then
                    components.Dispose()
                End If

            End If

            MyBase.Dispose(disposing)
        End Sub
#Region "Windows Form Designer generated code"

        Private Sub InitializeComponent()
            Dim resources As ComponentResourceManager = New ComponentResourceManager(GetType(WMTSForm))
            lServers = New Label()
            cbxServers = New ComboBox()
            btnConnect = New Button()
            lLayers = New Label()
            cbxLayers = New ComboBox()
            cbInvertAxis = New CheckBox()
            btnAdd = New Button()
            SuspendLayout()
            ' 
            ' lServers
            ' 
            lServers.AutoSize = True
            lServers.Location = New Point(18, 14)
            lServers.Margin = New Padding(4, 0, 4, 0)
            lServers.Name = "lServers"
            lServers.Size = New Size(69, 25)
            lServers.TabIndex = 0
            lServers.Text = "Servers"
            ' 
            ' cbxServers
            ' 
            cbxServers.Anchor = AnchorStyles.Top Or AnchorStyles.Left Or AnchorStyles.Right
            cbxServers.FormattingEnabled = True
            cbxServers.Items.AddRange(New Object() {"http://basemap.nationalmap.gov/arcgis/rest/services/USGSImageryOnly/MapServer/wmts", "http://garden.gis.vt.edu/arcgis/rest/services/VBMP2011/VBMP2011_Infrared_WGS/MapServer/WMTS/1.0.0/WMTSCapabilities.xml", "http://gis.oregonmetro.gov/services/wmts/1.0.0/WMTSGetCapabilities.xml", "http://maps.columbus.gov/arcgis/rest/services/Imagery/Imagery2013/MapServer/WMTS/1.0.0/WMTSCapabilities.xml", "https://mapy.geoportal.gov.pl/wss/service/PZGIK/ORTO/WMTS/StandardResolution"})
            cbxServers.Location = New Point(22, 38)
            cbxServers.Margin = New Padding(4, 4, 4, 4)
            cbxServers.Name = "cbxServers"
            cbxServers.Size = New Size(1104, 33)
            cbxServers.TabIndex = 1
            ' 
            ' btnConnect
            ' 
            btnConnect.Anchor = AnchorStyles.Top Or AnchorStyles.Right
            btnConnect.Location = New Point(1137, 34)
            btnConnect.Margin = New Padding(4, 4, 4, 4)
            btnConnect.Name = "btnConnect"
            btnConnect.Size = New Size(112, 34)
            btnConnect.TabIndex = 2
            btnConnect.Text = "Connect"
            btnConnect.UseVisualStyleBackColor = True
            ' 
            ' lLayers
            ' 
            lLayers.AutoSize = True
            lLayers.Location = New Point(18, 74)
            lLayers.Margin = New Padding(4, 0, 4, 0)
            lLayers.Name = "lLayers"
            lLayers.Size = New Size(61, 25)
            lLayers.TabIndex = 3
            lLayers.Text = "Layers"
            ' 
            ' cbxLayers
            ' 
            cbxLayers.Anchor = AnchorStyles.Top Or AnchorStyles.Left Or AnchorStyles.Right
            cbxLayers.FormattingEnabled = True
            cbxLayers.Location = New Point(22, 98)
            cbxLayers.Margin = New Padding(4, 4, 4, 4)
            cbxLayers.Name = "cbxLayers"
            cbxLayers.Size = New Size(974, 33)
            cbxLayers.TabIndex = 4
            ' 
            ' cbInvertAxis
            ' 
            cbInvertAxis.Anchor = AnchorStyles.Top Or AnchorStyles.Right
            cbInvertAxis.AutoSize = True
            cbInvertAxis.Location = New Point(1002, 100)
            cbInvertAxis.Margin = New Padding(4, 4, 4, 4)
            cbInvertAxis.Name = "cbInvertAxis"
            cbInvertAxis.Size = New Size(117, 29)
            cbInvertAxis.TabIndex = 5
            cbInvertAxis.Text = "Invert axis"
            cbInvertAxis.UseVisualStyleBackColor = True
            ' 
            ' btnAdd
            ' 
            btnAdd.Anchor = AnchorStyles.Top Or AnchorStyles.Right
            btnAdd.Location = New Point(1137, 94)
            btnAdd.Margin = New Padding(4, 4, 4, 4)
            btnAdd.Name = "btnAdd"
            btnAdd.Size = New Size(112, 34)
            btnAdd.TabIndex = 6
            btnAdd.Text = "Add"
            btnAdd.UseVisualStyleBackColor = True
            ' 
            ' WMTSForm
            ' 
            AutoScaleDimensions = New SizeF(144.0F, 144.0F)
            AutoScaleMode = AutoScaleMode.Dpi
            ClientSize = New Size(1286, 165)
            Controls.Add(btnAdd)
            Controls.Add(cbInvertAxis)
            Controls.Add(cbxLayers)
            Controls.Add(lLayers)
            Controls.Add(btnConnect)
            Controls.Add(cbxServers)
            Controls.Add(lServers)
            Icon = CType(resources.GetObject("$this.Icon"), Icon)
            Location = New Point(200, 120)
            Margin = New Padding(4, 4, 4, 4)
            Name = "WMTSForm"
            StartPosition = FormStartPosition.CenterScreen
            Text = "TatukGIS Samples"
            ResumeLayout(False)
            PerformLayout()

        End Sub
#End Region

        <STAThread()>
        Private Shared Sub Run()
            Application.Run(New WMTSForm)
        End Sub

        Private Sub WMTSForm_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load

        End Sub

        Private Sub btnConnect_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnConnect.Click
            Dim wmts As TGIS_LayerWMTS
            Dim lst As TList(Of TGIS_LayerInfo)
            wmts = New TGIS_LayerWMTS
            wmts.Path = cbxServers.Text
            Try
                lst = wmts.GetAvailableLayers
                For Each li As TGIS_LayerInfo In lst
                    cbxLayers.Items.Add(li.Name)
                Next
                If (cbxLayers.Items.Count > 0) Then
                    cbxLayers.SelectedIndex = 0
                End If
            Catch ex As EGIS_Exception
                MessageBox.Show(ex.Message)
            End Try

        End Sub

        Private Sub btnAdd_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnAdd.Click
            Dim wmts As TGIS_LayerWMTS
            Dim layer As TStrings
            Dim str As String
            Dim c() As Char
            wmts = New TGIS_LayerWMTS
            tkn = New TGIS_Tokenizer
            str = cbxLayers.Text
            c = New Char((1) - 1) {}
            c(0) = Microsoft.VisualBasic.ChrW(59)
            tkn.Execute(str, c)
            layer = tkn.Result
            wmts.Path = "[TatukGIS Layer\n" +
                        "Storage=WMTS\n" +
                        "Layer=" + layer(0) + "\n" +
                        "Url=" + cbxServers.Text + "\n" +
                        "TileMatrixSet=" + layer(2) + "\n" +
                        "ImageFormat=" + layer(1) + "\n" +
                        "InvertAxis=" + cbInvertAxis.Checked.ToString() + "\n"
            GIS.Add(wmts)
            If (GIS.Items.Count = 1) Then
                GIS.FullExtent()
            Else
                GIS.InvalidateWholeMap()
            End If

        End Sub
    End Class
End Namespace

File: WinForms/C#/WMTSManager/WMTSForm.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using TatukGIS.NDK;
using TatukGIS.NDK.WinForms;
using TatukGIS.RTL;

namespace WMTSManager
{
    /// <summary>
    /// Summary description for WMTSForm.
    /// </summary>
    public class WMTSForm : System.Windows.Forms.Form
    {
        private Label lServers;
        private ComboBox cbxServers;
        private Button btnConnect;
        private Label lLayers;
        private ComboBox cbxLayers;
        private CheckBox cbInvertAxis;
        private Button btnAdd;
        private TGIS_Tokenizer tkn;
        private TGIS_ViewerWnd GIS;

        public TGIS_ViewerWnd getGIS()
        {
            return GIS;
        }

        public void setGIS(TGIS_ViewerWnd _gis)
        {
            GIS = _gis;
        }

        /// <summary>
        /// Required designer variable.
        /// </summary>
        private System.ComponentModel.IContainer components;

        public WMTSForm()
        {
            //
            // Required for Windows Form Designer support
            //
            InitializeComponent();

            //
            // TODO: Add any constructor code after InitializeComponent call
            //
        }

        /// <summary>
        /// Clean up any resources being used.
        /// </summary>
        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                if (components != null)
                {
                    components.Dispose();
                }
            }
            base.Dispose(disposing);
        }

        #region Windows Form Designer generated code
        /// <summary>
        /// Required method for Designer support - do not modify
        /// the contents of this method with the code editor.
        /// </summary>
        private void InitializeComponent()
        {
            ComponentResourceManager resources = new ComponentResourceManager(typeof(WMTSForm));
            lServers = new Label();
            cbxServers = new ComboBox();
            btnConnect = new Button();
            lLayers = new Label();
            cbxLayers = new ComboBox();
            cbInvertAxis = new CheckBox();
            btnAdd = new Button();
            SuspendLayout();
            // 
            // lServers
            // 
            lServers.AutoSize = true;
            lServers.Location = new Point(18, 14);
            lServers.Margin = new Padding(4, 0, 4, 0);
            lServers.Name = "lServers";
            lServers.Size = new Size(69, 25);
            lServers.TabIndex = 0;
            lServers.Text = "Servers";
            // 
            // cbxServers
            // 
            cbxServers.Anchor = AnchorStyles.Top | AnchorStyles.Left | AnchorStyles.Right;
            cbxServers.FormattingEnabled = true;
            cbxServers.Items.AddRange(new object[] { "http://basemap.nationalmap.gov/arcgis/rest/services/USGSImageryOnly/MapServer/wmts", "http://garden.gis.vt.edu/arcgis/rest/services/VBMP2011/VBMP2011_Infrared_WGS/MapServer/WMTS/1.0.0/WMTSCapabilities.xml", "http://gis.oregonmetro.gov/services/wmts/1.0.0/WMTSGetCapabilities.xml", "http://maps.columbus.gov/arcgis/rest/services/Imagery/Imagery2013/MapServer/WMTS/1.0.0/WMTSCapabilities.xml", "https://mapy.geoportal.gov.pl/wss/service/PZGIK/ORTO/WMTS/StandardResolution" });
            cbxServers.Location = new Point(22, 38);
            cbxServers.Margin = new Padding(4, 4, 4, 4);
            cbxServers.Name = "cbxServers";
            cbxServers.Size = new Size(1104, 33);
            cbxServers.TabIndex = 1;
            // 
            // btnConnect
            // 
            btnConnect.Anchor = AnchorStyles.Top | AnchorStyles.Right;
            btnConnect.Location = new Point(1137, 34);
            btnConnect.Margin = new Padding(4, 4, 4, 4);
            btnConnect.Name = "btnConnect";
            btnConnect.Size = new Size(112, 34);
            btnConnect.TabIndex = 2;
            btnConnect.Text = "Connect";
            btnConnect.UseVisualStyleBackColor = true;
            btnConnect.Click += btnConnect_Click;
            // 
            // lLayers
            // 
            lLayers.AutoSize = true;
            lLayers.Location = new Point(18, 74);
            lLayers.Margin = new Padding(4, 0, 4, 0);
            lLayers.Name = "lLayers";
            lLayers.Size = new Size(61, 25);
            lLayers.TabIndex = 3;
            lLayers.Text = "Layers";
            // 
            // cbxLayers
            // 
            cbxLayers.Anchor = AnchorStyles.Top | AnchorStyles.Left | AnchorStyles.Right;
            cbxLayers.FormattingEnabled = true;
            cbxLayers.Location = new Point(22, 98);
            cbxLayers.Margin = new Padding(4, 4, 4, 4);
            cbxLayers.Name = "cbxLayers";
            cbxLayers.Size = new Size(974, 33);
            cbxLayers.TabIndex = 4;
            // 
            // cbInvertAxis
            // 
            cbInvertAxis.Anchor = AnchorStyles.Top | AnchorStyles.Right;
            cbInvertAxis.AutoSize = true;
            cbInvertAxis.Location = new Point(1002, 100);
            cbInvertAxis.Margin = new Padding(4, 4, 4, 4);
            cbInvertAxis.Name = "cbInvertAxis";
            cbInvertAxis.Size = new Size(117, 29);
            cbInvertAxis.TabIndex = 5;
            cbInvertAxis.Text = "Invert axis";
            cbInvertAxis.UseVisualStyleBackColor = true;
            // 
            // btnAdd
            // 
            btnAdd.Anchor = AnchorStyles.Top | AnchorStyles.Right;
            btnAdd.Location = new Point(1137, 94);
            btnAdd.Margin = new Padding(4, 4, 4, 4);
            btnAdd.Name = "btnAdd";
            btnAdd.Size = new Size(112, 34);
            btnAdd.TabIndex = 6;
            btnAdd.Text = "Add";
            btnAdd.UseVisualStyleBackColor = true;
            btnAdd.Click += btnAdd_Click;
            // 
            // WMTSForm
            // 
            AutoScaleDimensions = new SizeF(144F, 144F);
            AutoScaleMode = AutoScaleMode.Dpi;
            ClientSize = new Size(1286, 165);
            Controls.Add(btnAdd);
            Controls.Add(cbInvertAxis);
            Controls.Add(cbxLayers);
            Controls.Add(lLayers);
            Controls.Add(btnConnect);
            Controls.Add(cbxServers);
            Controls.Add(lServers);
            Icon = (Icon)resources.GetObject("$this.Icon");
            Location = new Point(200, 120);
            Margin = new Padding(4, 4, 4, 4);
            Name = "WMTSForm";
            StartPosition = FormStartPosition.CenterScreen;
            Text = "TatukGIS Samples";
            Load += WMTSForm_Load;
            ResumeLayout(false);
            PerformLayout();

        }
        #endregion

        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        [STAThread]
        static void Run()
        {
            Application.Run(new WMTSForm());
        }

        private void WMTSForm_Load(object sender, System.EventArgs e)
        {
        }

        private void btnConnect_Click(object sender, EventArgs e)
        {
            TGIS_LayerWMTS wmts;
            TList<TGIS_LayerInfo> lst;

            wmts = new TGIS_LayerWMTS();
            wmts.Path = cbxServers.Text;
            try {
                lst = wmts.GetAvailableLayers();
                foreach (TGIS_LayerInfo li in lst)
                    cbxLayers.Items.Add(li.Name);
                if (cbxLayers.Items.Count > 0)
                    cbxLayers.SelectedIndex = 0;
            }
            catch ( EGIS_Exception ex)
            {
                MessageBox.Show(ex.Message);
            }
        }

        private void btnAdd_Click(object sender, EventArgs e)
        {
            TGIS_LayerWMTS wmts;
            TStrings layer;
            String str;
            char[] c;

            wmts = new TGIS_LayerWMTS();
            tkn = new TGIS_Tokenizer();

            str = cbxLayers.Text;
            c = new char[1];
            c[0] = ';';
            tkn.Execute(str, c);

            layer = tkn.Result;

            wmts.Path = "[TatukGIS Layer\n" +
                        "Storage=WMTS\n" +
                        "Layer=" + layer[0] + "\n" +
                        "Url=" + cbxServers.Text + "\n" +
                        "TileMatrixSet=" + layer[2] + "\n" +
                        "ImageFormat=" + layer[1] + "\n" +
                        "InvertAxis=" + cbInvertAxis.Checked.ToString() + "\n";

            GIS.Add(wmts);
            if (GIS.Items.Count == 1)
                GIS.FullExtent();
            else
                GIS.InvalidateWholeMap();

        }
    }
}
